internal/httpclient/registry_client.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package httpclient
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+// NewForRegistryRequests is a variant of [New] that deals with some additional
+// policy concerns related to "registry requests".
+//
+// Exactly what constitutes a "registry request" is actually more about
+// historical technical debt than intentional design, since these concerns
+// were originally supposed to be handled internally within the module and
+// provider registry clients but the implementation of that unfortunately caused
+// the effects to "leak out" into other parts of the system, which we now
+// preserve for backward compatibility here.
+//
+// Therefore "registry requests" includes the following:
```

Comment:
This comment was super helpful to understand why we do this for disco and for module/provider clients. thanks!